File: 139_Word_Break/139_1.py
from typing import List

class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        wordSet = set(wordDict)
        size = len(s)
        memo = [None] * size
        def dfs(start) -> bool:
            if start >= size:
                return True
            if memo[start] is not None:
                return memo[start]
            for j in range(start + 1, size + 1):
                if s[start: j] in wordSet and dfs(j):
                    memo[start] = True
                    return True
            memo[start] = False
            return False
        return dfs(0)

File: 139_Word_Break/139_1.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        int len = s.length();
        Boolean[] visited = new Boolean[len];
        Arrays.fill(visited, false);
        Set<String> wordSet = new HashSet<>(wordDict);
        Deque<Integer> queue = new ArrayDeque<>();
        queue.offerLast(0);
        while (!queue.isEmpty()) {
            int cur = queue.removeFirst();
            if (visited[cur]) continue;
            visited[cur] = true;
            for (int i = cur + 1; i <= len; i++) {
                String substr = s.substring(cur, i);
                if (wordSet.contains(substr)) {
                    if (i < len) {
                        queue.offerLast(i);
                    } else {
                        return true;
                    }
                }
            }
        }
        return false;
    }
}

public class MainClass {
    public static void main(String[] args) {
        Solution s = new Solution();
        List<String> wordDict = new ArrayList<>();
        wordDict.add("leet");
        wordDict.add("code");
        System.out.println(s.wordBreak("leetcode", wordDict));
    }
}
